src/System.Private.CoreLib/shared/System/Numerics/Vector4_Intrinsics.cs

```diff
@@ -150,10 +180,35 @@ public Vector4(Vector3 value, float w)
         /// <param name="vector1">The first vector.</param>
         /// <param name="vector2">The second vector.</param>
         /// <returns>The dot product.</returns>
+#if ARM64
         [Intrinsic]
+#endif
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static float Dot(Vector4 vector1, Vector4 vector2)
         {
+            if (Sse41.IsSupported)
+            {
+                return Sse41.DotProduct(vector1.AsVector128(), vector2.AsVector128(), 0xFF).ToScalar();
+            }
+            else if (Sse.IsSupported)
+            {
+                Vector128<float> tmp = Sse.Multiply(vector1.AsVector128(), vector2.AsVector128());
+
+                if (Sse3.IsSupported)
+                {
+                    tmp = Sse3.HorizontalAdd(tmp, tmp);
+                    return Sse3.HorizontalAdd(tmp, tmp).ToScalar();
+                }
+                else
+                {
+                    Vector128<float> tmp2 = Sse.Shuffle(vector2.AsVector128(), tmp, 0x40);
+                    tmp2 = Sse.Add(tmp2, tmp);
+                    tmp = Sse.Shuffle(tmp, tmp2, 0x30);
```

Comment:
In the absence of the MM_SHUFFLE macro, I find binary literals are almost as easy to read, and certainly easier than hex.  e.g. 0b_00_11_00_00 here

Reply:
We should really just add a helper intrinsic for this, it helps readability immensely and is going to be used in most native code people might try to port.

I don't think we have an API proposal yet, however.

Reply:
I've been waiting for the _MM_SHUFFLE API since the S.R.I beginning 🙂 I guess it should somehow only allow constant values as an input.

Reply:
> I guess it should somehow only allow constant values as an input.

There is currently no C# language-specific way to enforce it and the only way is via analyzers.

There is a C# proposal for runtime `const unmanaged` but despite preliminary OK from language team it seems that interest has faded (see https://github.com/dotnet/csharplang/issues/744)